### PR TITLE
feat(spell): use hunspell as the spellchecking backend

### PR DESCRIPTION
**Pull request by vigoux**
# TLDR

This PR enables the use of `hunspell`.

If you want to help:
- Try this PR _in your language_
- Report any error (they are probably caused by encoding problems).

# Mandatory motivation / problem disclaimer

Spellchecking in vim is notoriously bad at handling anything but english.
Furthermore, spelling suggestions are known to take a very long time to run (acknowledged in the docs).

Note though that it does not integrate within system dictionnaries by default (hunspell as a library does not enable that), but can use legacy vim spell files (the `.dic` and `.aff` files, but not the `.spl` ones that are specific to vim).

# Why hunspell ?

`hunspell` seems to be the most mature spellchecking project. It is supported on every single platform we can think of, and provides UTF-8 handling by default.

`hunspell` seems easier to include in our codebase, as it uses iconv, which we already include in out codebase.

In terms of performances, hunspell seems equivalent to vimspell when checking, and much faster when suggesting.

Other alternatives:
- `enchant`: requires `GLib`
- `nuspell`: misses some things like `SpellRare` support, requires `ICU`.

# Status
- Hunspell itself and dictionnaries
  - [x] Add `*.add` dictionnaries at runtime
  - [ ] `spellfile` (a bit tricky, might require manual parsing of the file...)
  - [ ] Add `spellpath` option for the list of paths to search for dictionnaries
- Spellchecking functions
  - [x] `spellsuggest()`
  - [x] `spellbadword()`
  - [x] `vim.spell.check()`
- Spell mappings
  - [x] `z=`
  - [x] `zg`
  - [x] `]s` `[s`
  - [x] `zG`
  - [ ] `zw`
  - [x] `zW`
- Spell commands
  - [x] `spellgood`
  - [x] `spellundo`
  - [x] `spellinfo`
  - [ ] `spellbad`
  - [ ] others
- Misc.:
  - [x] `<C-X><C-S>`

Things that are removed because handled by hunspell:
- Suggestion options:
  - `spellsuggest=double`
- Sound-folding in general (and `soundfold()`)
- `:mkspell`

Other marginal improvements:
- The internal wordlist is now stored in memory, instead of on-disk

### CODE REVIEW

**Comment by clason** [comment]
If you don't bundle hunspell (which is OK for an optional feature), you need to add it to the build image here:
https://github.com/neovim/neovim/blob/22a88c425e2fe99bcb534e43396a0b00044ae67d/.github/workflows/ci.yml#L36

**Comment by vigoux** [comment]
Will do :+1:

File: src/hunspell/hunspell_wrapper.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#ifndef __HUNSPELL_WRAPPER_H__
+#define __HUNSPELL_WRAPPER_H__
+
```
Comment (ii14) [inline src/hunspell/hunspell_wrapper.h:3]:
```suggestion
#ifdef __cplusplus
extern "C" {
#endif
```

File: src/hunspell/hunspell_wrapper.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+typedef struct hunspell_S hunspell_T;
+
+hunspell_T * hunspell_create(const char *affpath, const char *dicpath);
+
+void hunspell_destroy(hunspell_T *pHunspell);
+
+bool hunspell_is_wordchar(hunspell_T *handle, const char *p);
+
+bool hunspell_spell_flags(hunspell_T *handle, const char *p, size_t len, int *flags);
```
Comment (ii14) [inline src/hunspell/hunspell_wrapper.h:24]:
```suggestion
bool hunspell_spell_flags(hunspell_T *handle, const char *p, size_t len, int *flags);

#ifdef __cplusplus
}
#endif
```

File: src/hunspell/hunspell_wrapper.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#include <hunspell/hunspell.hxx>
+#include <hunspell/hunspell.h>
+
+using std::string;
+
+extern "C" {
```
Comment (ii14) [inline src/hunspell/hunspell_wrapper.cpp:6]:
```suggestion
```

File: src/hunspell/hunspell_wrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+bool hunspell_spell_flags(hunspell_T* handle, const char *p, size_t len, int *flags)
+{
+  string tospell(p);
+  Hunspell *h = reinterpret_cast<Hunspell*>(handle);
+
+  bool ret = h->spell(tospell, flags);
+
+  return ret;
+}
+}
```
Comment (ii14) [inline src/hunspell/hunspell_wrapper.cpp:39]:
```suggestion
```

File: cmake/FindHunspell.cmake [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# - Try to find tree-sitter
```
Comment (jamessan) [inline cmake/FindHunspell.cmake:1]:
```suggestion
# - Try to find Hunspell
```

Comment (clason) [reply cmake/FindHunspell.cmake:1]: and similarly for the other copypasta below ;)

File: src/hunspell/hunspell_wrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
+#include <hunspell/hunspell.hxx>
+#include <hunspell/hunspell.h>
+#include <string.h>
```
Comment (ii14) [inline src/hunspell/hunspell_wrapper.cpp:3]:
```suggestion
#include <cstring>
```

File: src/hunspell/hunspell_wrapper.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#ifndef __HUNSPELL_WRAPPER_H__
+#define __HUNSPELL_WRAPPER_H__
```
Comment (ii14) [inline src/hunspell/hunspell_wrapper.h:2]: Double underscores are reserved, it's best to avoid them
```suggestion
#ifndef HUNSPELL_WRAPPER_H
#define HUNSPELL_WRAPPER_H
```

**Comment by vigoux** [comment]
So I think this is pretty much working (but with some really rough edges).

I'll now work on integrating better all this, like correctly handle the changes of `spelloptions` to clear the language list and such.

File: nvim_hunspell/builtin.lua [hunk 8]
```diff
@@ -0,0 +1,39 @@
+vim.o.spelloptions = "hunspell"
+vim.o.spelllang = "en"
+vim.o.spell = true
+
+local total_time = os.time()
```
Comment (lewis6991) [inline nvim_hunspell/builtin.lua:5]: Better use `vim.loop.hrtime()` for benchmarking.

File: nvim_hunspell/builtin.lua [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  errors = 0
+  for line in function() return file:read "*l" end do
+    local bad, suggs_plain = unpack(vim.split(line, "\t", { plain = true }))
+    print(bad)
+
+    local suggs = vim.split(suggs_plain, ', ', { plain = true, trimempty = true })
+
+    local bad_spell = vim.fn.spellbadword(bad)[1]
```
Comment (lewis6991) [inline nvim_hunspell/builtin.lua:22]: Don't forget we also have `vim.spell.check` which avoids the viml bridge.

File: nvim_hunspell/builtin.lua (hunk 8 above)
Comment (vigoux) [reply nvim_hunspell/builtin.lua:5]: Will do then!

File: nvim_hunspell/builtin.lua (hunk 9 above)
Comment (vigoux) [reply nvim_hunspell/builtin.lua:22]: Oh I forgot about that, I'll try it this way then.

**Comment by vigoux** [comment]
Updated benchmark timings for `spell.check`:
vimSpell: `17159` ms
hunspell: `21944` ms

Note though that the two do not use the same dictionnary.

**Comment by clason** [comment]
And the _real_ test will be suggestions anyway, especially for morphologically complex languages like French and German -- here vimSpell is notoriously bad, to the point where people often want to cancel out, if only they could...

**Review by justinmk — CHANGES_REQUESTED** [review]
Where is the "Problem" (or "Motivation") statement (max 500 chars, not a wall of text)? That is a required first step for all non-trivial PRs. Note that "I want to use the same saved words in all of my applications" is not a strong enough motivation. 

Related (wall of text, does not explain in 500 chars the motivation for this PR): https://github.com/neovim/neovim/issues/12064


Everything we add to the project has a cost. 

- We've made compromises for treesitter because it gains tons of value. 
- We added cjson because (1) it's performance-critical, (2) it was rather isolated, and (3) may potentially eliminate the vim json implementation in the future.
- We added xdiff because vim did it and it's isolated.

The code in this PR is not isolated, it appears to need many changes in the core spell modules. Will this make maintenance harder when it comes to spell-related bugs? AFAIK, we have not inlined any C++ code and I am very reluctant to open that door.

# Path forward

This would be compelling if all of these are true:

- ❓ Eliminate the inlined c++ code. We shouldn't depend on a c++ compiler (for non-bundled deps), that is a new dependency.
    - Link to hunspell from system (or bundled) libs.
    - Make hunspell a required dep. "Optional build-time features" are a complication that we have avoided for a very long time (except for very limited cases).
    - Specify that we choose the LGPL license of hunspell.
- ❓ Hunspell has significantly better and well-maintained dictionaries compared to vim spell.
    - vimspell works poorly for non-English/Dutch languages.
    - hunspell's suggest algorithm is worse for English...
- ✅ Users on all platforms have a way to install dictionaries (compare how currently Nvim auto-downloads spellfiles).
    - We could host them in https://github.com/neovim/deps perhaps.
- ✅ **Remove support for the old vim spellfiles**.
    - Accept the limitations of hunspell and remove features that hunspell can't support.
    - This allows us to focus on one paradigm, to keep maintenance burden manageable.

**Comment by jamessan** [comment]
> * Hunspell has significantly better and well-maintained dictionaries compared to vim spell.
> 
> * Nvim automatically pulls dictionaries somehow (similar to the old vim dictionaries).
>   
>   * We could host them in [neovim/deps](https://github.com/neovim/deps) perhaps.

Vim's spellfiles are essentially unmaintained.  That, to me, is the big win of using an external spell library.  Offload the maintenance (including distribution of the dictionaries) to the developers of the spell library.

It has nothing to do with "use the same saved words in all of my applications", but if that happens to be part of the end result, even better.

**Comment by mcepl** [comment]
> Vim's spellfiles are essentially unmaintained. That, to me, is the big win of using an external spell library. Offload the maintenance (including distribution of the dictionaries) to the developers of the spell library.
> 
> It has nothing to do with "use the same saved words in all of my applications", but if that happens to be part of the end result, even better.

Also, having just one user spell dictionary for whole system, so I don’t have to put same words to two different user dictionaries, is a plus.

**Comment by mcepl** [comment]
Hmm, trying to put the patch from this PR on the top of the current master and I get this failure to build:
```
[   48s] [100%] Building C object src/nvim/CMakeFiles/nvim.dir/normal.c.o
[   48s] cd "/home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/build/src/nvim" && /usr/bin/cc -DINCLUDE_GENERATED_DECLARATIONS -DMIN_LOG_LEVEL=3 -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -D_GNU_SOURCE -Dnvim_EXPORTS -I"/home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/build/cmake.config" -I"/home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src" -I"/home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/build/src/nvim/auto" -I"/home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/build/include" -I/usr/include/luajit-5_1-2.1 -isystem /usr/include/lua5.1 -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -fcommon -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR  -Og -g -flto=auto -fno-fat-lto-objects -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -MD -MT src/nvim/CMakeFiles/nvim.dir/normal.c.o -MF CMakeFiles/nvim.dir/normal.c.o.d -o CMakeFiles/nvim.dir/normal.c.o -c "/home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c"
[   48s] /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c:85:1: error: conflicting types for 'get_visual_text'; have '_Bool(cmdarg_T *, char_u **, size_t *)' {aka '_Bool(struct cmdarg_S *, unsigned char **, long unsigned int *)'}
[   48s]    85 | get_visual_text (
[   48s]       | ^~~~~~~~~~~~~~~
[   48s] In file included from /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.h:76,
[   48s]                  from /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/ex_cmds_defs.h:8,
[   48s]                  from /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/vim.h:265,
[   48s]                  from /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/getchar.h:5,
[   48s]                  from /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/api/private/helpers.h:7,
[   48s]                  from /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c:16:
[   48s] /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/build/include/normal.h.generated.h:35:17: note: previous declaration of 'get_visual_text' with type '_Bool(cmdarg_T *, char **, size_t *)' {aka '_Bool(struct cmdarg_S *, char **, long unsigned int *)'}
[   48s]    35 | DLLEXPORT _Bool get_visual_text(cmdarg_T *cap, char **pp, size_t *lenp);
[   48s]       |                 ^~~~~~~~~~~~~~~
[   48s] /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c: In function 'nv_zg_zw':
[   48s] /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c:3585:46: warning: passing argument 2 of 'get_visual_text' from incompatible pointer type [-Wincompatible-pointer-types]
[   48s]  3585 |   if (VIsual_active && !get_visual_text(cap, &ptr, &len)) {
[   48s]       |                                              ^~~~
[   48s]       |                                              |
[   48s]       |                                              char **
[   48s] /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c:87:14: note: expected 'char_u **' {aka 'unsigned char **'} but argument is of type 'char **'
[   48s]    87 |     char_u **pp,
[   48s]       |     ~~~~~~~~~^~
[   48s] /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c: In function 'nv_ident':
[   48s] /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c:4274:47: warning: passing argument 2 of 'get_visual_text' from incompatible pointer type [-Wincompatible-pointer-types]
[   48s]  4274 |     if (VIsual_active && get_visual_text(cap, &ptr, &n) == false) {
[   48s]       |                                               ^~~~
[   48s]       |                                               |
[   48s]       |                                               char **
[   48s] /home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/src/nvim/normal.c:87:14: note: expected 'char_u **' {aka 'unsigned char **'} but argument is of type 'char **'
[   48s]    87 |     char_u **pp,
[   48s]       |     ~~~~~~~~~^~
[   48s] make[2]: *** [src/nvim/CMakeFiles/nvim.dir/build.make:3274: src/nvim/CMakeFiles/nvim.dir/normal.c.o] Error 1
[   48s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/neovim-0.8.0~git.1662905274.f98cff957/build'
```

**Comment by vigoux** [comment]
I added a disclaimer in the description of the PR.


> * ballot_box_with_check  Eliminate the inlined c++ code. We shouldn't depend on a c++ compiler (for non-bundled deps), that is a new dependency.
>   
>   * Link to hunspell from system (or bundled) libs.
>   * Make hunspell a required dep. "Optional build-time features" are a complication that we have avoided for a very long time (except for very limited cases).
>   * Specify that we choose the LGPL license of hunspell.

I'll develop more about this when answering to the last point, but using the C interface to hunspell would prevent us from getting more detailed informations out of it.

I'll look into this and see what I can do anyway.

> * white_check_mark  Hunspell has significantly better and well-maintained dictionaries compared to vim spell.
> * ballot_box_with_check  Users on all platforms have a way to install dictionaries (compare how currently Nvim auto-downloads spellfiles).
>   
>   * We could host them in https://github.com/neovim/deps perhaps.

These three point are valid and I don't see if it blocks adoptance of hunspell in any way.

> * ballot_box_with_check **Remove support for the old vim spellfiles**.
>   
>   * Accept the limitations of hunspell and remove features that hunspell can't support.
>   * This allows us to focus on one paradigm, to keep maintenance burden manageable.

This is the interesting bit. When discussing about integrating hunspell, it was said that we should keep backward compatibility.

Not keeping support for legacy vim spellfiles would greatly simplify the code, and would maybe allow a deeper rework of the whole spellchecking codebase.

The amount of work would quite probably be more significant, but as your said, it would quite probably lower the maintenance burden a lot.

This point also has to be put in perspective with the first (C++ dependency). We could use the C-only API for hunspell, but writing our own wrapper for C++ APIs seems better as the C APIs are not really ideal (as an example, spellchecking in C++ allows to get details about the word being spellchecked, while that's not possible in C).

While I get that this is a problem with regard to the amount of dependency this PR adds, I this that the overrall advantages are numerous:
- Better performances
- Better language support
- Outsourcing dictionnaries / codebase management
- Lowering the development barrier of entry with regard to spell features.

**Comment by justinmk** [comment]
> This is the interesting bit. When discussing about integrating hunspell, it was said that we should keep backward compatibility.

Unless hunspell is missing major features of vimspell, I think we need to just pick one. If we start maintaining a hunspell interface, we're in the hunspell business and we are no longer gaining much from supporting vimspell. At that point, vimsplle is just a burden. We need to pick one and commit to it.

**Comment by lewis6991** [comment]
Other than vim patches, I don't think vimspell has required any real maintenance in years.

There needs to be some overlap of both existing for compat and user migration before we can consider choosing only one.

**Comment by justinmk** [comment]
> Other than vim patches, I don't think vimspell has required any real maintenance in years (if ever).

That will change if we have two subsystems, because both must constantly be reconciled with the other.

> There needs to be some overlap of both existing for compat and user migration before we can consider choosing only one.

What is a concrete example? I expect the same family of spell-related functions will mostly work, as well as mappings. So I don't understand where "migration" is needed.

**Comment by lewis6991** [comment]
> That will change if we have two subsystems, because both must constantly be reconciled with the other.

If vimspell isn't being actively developed, then I whole heartedly disagree this will introduce any additional burden. If it does actually introduce burden that we aren't happy with then we can make the decision to remove it then, as opposed to speculating.

> What is a concrete example? I expect the same family of spell-related functions will mostly work, as well as mappings. So I don't understand where "migration" is needed.

Fair. I guess if hunspell works similarly enough then this won't be a problem. We just don't know that yet, so having a workaround to offer users is preferable, especially while we work out the finer details, and especially since the perceived cost (from my POV) is quite low.

Bram spent a long time tuning and optimizing vimspell for English, so I wouldn't be surprised if it is actually better for English. I would definitely like to find out first.

**Comment by vigoux** [comment]
(drive by comment but I wanted to answer some points)

Mainly regarding optimization and vim spell for english: the fun part is that vim spell is both worse and slower (for the suggestion part) than hunspell.

Bram himself admitted it: the code he wrote back then was to work around the lack of good libraries, but now they are probably okay.

There are a couple of good points that come with hunspell integration:
- drop-in dictionaries, maintained elsewhere
- actual support for non-English languages (I am speaking of features like compound words)
- better performances
- no weird compilation overhead for spell files
- (maybe) no need to download spell files from some FTP server: we could allow setting a search path for dictionnaries so that packagers allow to use the system dictionaries

**Comment by mcepl** [comment]
> Other than vim patches, I don't think vimspell has required any real maintenance in years (if ever).

I am not sure whether the word “required” is the best word to use. The truth is also that `spell.c` is the longest and most complicated file in the vim codebase and so nobody has dared to touch it.

**Comment by justinmk** [comment]
> If vimspell isn't being actively developed, then I whole heartedly disagree this will introduce any additional burden. If it does actually introduce burden that we aren't happy with then we can make the decision to remove it then, as opposed to speculating.

As usual, there is at least an interface burden. This means documentation would be more complicated, not to mention internals. Avoiding a decision here is a cop-out. A key goal of Neovim is to eliminate this kind of thing.

**Comment by lewis6991** [comment]
In that case doesn't the exact same principle apply to treesitter?

**Comment by mcepl** [comment]
> As usual, there is at least an interface burden. This means documentation would be more complicated, not to mention internals. Avoiding a decision here is a cop-out. A key goal of Neovim is to eliminate this kind of thing.

I think that both for tree-sitter and this could be temporary solution until tree-sitter and external-spellchecker are mature enough to be the only implementation.

**Comment by mcepl** [comment]
> If hunspell turns out to be a mistake we can revert. Should be able to discover that before the next major release. And in that case I do not think we should support hunspell at all. Avoid kitchen-sink.

I think that even more important than switch particularly to hunspell is to untangle spelling in neovim to clean API, which then can be served by another external spellchecker. E.g., I have heard some rumours about https://nuspell.github.io/ (and there are of course system spellcheckers in Mac OS X and Windows), so the spellchecking engine which will be used is not so important as the creation of API for communication with the external library.

**Comment by vigoux** [comment]
If we are willing to convert the existing spellchecking code to another platform, it may be worth considering nuspell though.

Opening the door to another spellchecking library is nice, and maybe supporting a recent tool (like nuspell, which might bring some performance benefits even more than hunspell) might be a better long-term decision.

**Comment by mcepl** [comment]
> If we are willing to convert the existing spellchecking code to another platform, it may be worth considering nuspell though.

Well, so far I see on Linux platforms hunspell as the default system spellchecker (used by, among others, LibreOffice, Thunderbird, Chromium, and Firefox), but yes nuspell can use hunspell dictionaries, so it is not that important.

**Comment by vigoux** [comment]
Looking a little deeper into this, nuspell lacks some interesting features like helpers for tokenization of words, and finer-grained reports on the found word. 

I think hunspell is the safest bet here.

**Comment by vigoux** [comment]
Following discussion on matrix: this PR will shift its goal to __remove legacy vim spell__.
It will take a little more time than expected, but I expect the results to be much better.

**Comment by vigoux** [comment]
_UPDATE_: This PR now ships a default `en` dictionnary. One can thus test it with this dictionnary already.

**Comment by mcepl** [comment]
So, how safe this is to use? I had this PR in my build just to check whether it is buildable, but when vimspell is going to be removed, will it actually break my spellchecker? (And no, I don’t ask for 100% quality, just whether I get anything from my spellchecker now)

**Comment by justinmk** [comment]
What's the purpose of the new top-level dir `nvim_hunspell/` ? I would think those files should live in runtime/ somewhere. Perhaps `spell2/` could be where we look for hunspell dictionaries, but more likely we just need a recognizable filename extension: `spell/en.utf-8.hunspell` or whatever (hopefully hunspell already has a standard extension).

**Comment by clason** [comment]
Reminder: This is a draft PR for following along with the experimentation, not a PR ready to merge!

**Comment by vigoux** [comment]
> more likely we just need a recognizable filename extension: `spell/en.utf-8.hunspell` or whatever (hopefully hunspell already has a standard extension).

This is already the case as of this implementation:
- Vim only recognizes filenames like `{lang}.{encoding}.dic`
- This implementation finds names like `{lang}.dic`

**Comment by vigoux** [comment]
_PROPOSAL_: It could be nice that our `spellfile` format changes (maybe also allowing arbitrary names that do not collide with `.add` which are meant to be dictionnaries extensions), se that it maps the `man 5 hunspell` "Personal dictionary file" format.

Otherwise, we could just ditch the loading of `.add` files (which is the remaining of the autoloading of `.add.spl` files), and use only those specified in `spellfile`.

In any case, I think that `spellfile`, as a user-facing extension mechanism, should use a similar (if not _the same_) format than the one in hunspell.

**Comment by arp242** [comment]
> Unless hunspell is missing major features of vimspell, I think we need to just pick one. 

For English and Dutch at least I found that Vim's spell checking is consistently much better than hunspell. I regularly have to copy/paste words from Firefox (which uses hunspell) to Vim to get any meaningful suggestions at all, and the ordering is usually much better in Vim.

I believe aspell is often better than hunspell for English too. Not the entire world is English of course and no doubt the situation is different for other languages, but English is obviously a major use case.

---

I wonder if something like a `spellfunc` option might be a better path forward for integrating hunspell or other tools; you can then write a function to parse the hunspell CLI:

    % hunspell -d en_US <<<helo
    Hunspell 1.7.0
    & helo 11 0: hole, help, helot, hello, halo, hero, hell, held, helm, he lo, he-lo

Or to check an entire file:

    % hunspell -d en_US -a [file]

Should also be much easier to implement, although need to think about how to best implement :spellgood, :spellbad, etc.

This way you could even have a `vimspell` CLI, set the best spell check tool for the language you're currently using, etc.

**Comment by justinmk** [comment]
> In any case, I think that `spellfile`, as a user-facing extension mechanism, should use a similar (if not _the same_) format than the one in hunspell.

If we are using hunspell then the dict format should be the identical hunspell dict format. Otherwise we have to convert things, which leads to stale files, as seen with vimspell itself. It also adds a user burden (extra steps), complicates documentation, etc.

> something like a spellfunc option might be a better path forward ... then write a function to parse the hunspell CLI ... Should also be much easier to implement, although need to think about :spellgood, :spellbad, etc.

@arp242 Thanks for your feedback. The dealbreaker that led to deciding against 'spellfunc' is performance. If you want to try out this PR (or wait until it reaches mainline), comparing the results for English/Dutch will be valuable.

File: src/nvim/spell.c [hunk 10]
```diff
@@ -1539,20 +839,43 @@ static void spell_load_lang(char_u *lang)
...
+      // TODO(vigoux): convert to lua
       char autocmd_buf[512] = { 0 };
       snprintf(autocmd_buf, sizeof(autocmd_buf),
                "autocmd VimEnter * call spellfile#LoadFile('%s')|set spell",
                lang);
       do_cmdline_cmd(autocmd_buf);
     } else {
-      smsg(_("Warning: Cannot find word list \"%s.%s.spl\" or \"%s.ascii.spl\""),
-           lang, spell_enc(), lang);
+      smsg(_("Warning: Cannot find dictionnary for \"%s\""), lang);
```
Comment (arp242) [inline src/nvim/spell.c:849]: dictionary is misspelled here.

**Comment by arp242** [comment]
> The dealbreaker that led to deciding against 'spellfunc' is performance.

I haven't looked at it in detail (and I couldn't find any discussion it on the issue tracker), but `hunspell -l` on a 4,800 word document is ~70ms on my six-year old not-fast-at-all laptop, and `hunspell -d en_US <<<decompilation` to actually get the suggestions is ~5ms. A 83,594 word document is ~500ms though, which is a tad slow.

There is some added overhead from parsing and such, but that shouldn't be too much, and if hunspell-the-library can be fast enough, then I don't see why hunspell-the-CLI needs to be slow (there is probably something to win here).

> If you want to try out this PR (or wait until it reaches mainline), comparing the results for English/Dutch will be valuable.

`concencus` is `['consensus', 'con census', 'consents', 'concerns', 'concepts', 'consciences']` in Vim, and `['concurrences', 'concupiscence', 'concurrence', 'concussion']` with this PR ("consensus" what was intended, the Firefox suggestions are the same).

`Helo` is `['Hello', 'Halo', 'Hell', 'Hole', 'Hero', 'Held']` in Vim, and `['He lo', 'He-lo', 'Hole', 'Help', 'Helot', 'Hello']` with this patch. "Hello" is the "obvious" suggestion, but it's listed as fifth in hunspell (Firefox here is even more hopeless and doesn't list "Hello" at all).

There are some other pathological use cases with Hunspell; the "concencus" is something that I happened to encounter this week so I remembered it. I've heard some other people complain about it over the years too. I don't know if there are kobs to tweak in Hunspell or if anything can be improved, but Hunspell in Firefox has been a major PITA for me for years. Vim, but also just entering the word in DuckDuckGo or Google (entering "concencus" there also corrects it to the intended "consensus"), are better spell checkers for English than Hunspell.

**Comment by lewis6991** [comment]
> There is some added overhead from parsing and such, but that shouldn't be too much, and if hunspell-the-library can be fast enough, then I don't see why hunspell-the-CLI needs to be slow (there is probably something to win here).

The initial version of https://github.com/lewis6991/spellsitter.nvim in fact used the CLI version of hunspell and was very slow. This is because spellchecking is done whilst redrawing each line.  So for a roughly 80 line redraw the cost will be 80x5ms = 400ms, so half a second for every full screen redraw. I would guess most of this cost would be coming from the system call + the pipe IO.

Spellsitter did also cache the results. It helped, but not enough since you need to invalidate the cache on every `changetick`.

**Comment by lewis6991** [comment]
> Why does it need to run on every redraw?

That's just what it does now. If you don't do it when drawing lines, then you have to do it on file load for the whole buffer.

> And why does it need to send every line separately?

It doesn't need to, that's just what it does currently. This change only switches out the backend, it doesn't completely re-implement how spell checking is applied. The current implementation is quite naive/simple and re-implementing it to work around the bottlenecks of calling out to the system will be a huge amount of work, and not worth it considering these spellchecker backends are supposed to loaded as libraries anyway.

> Either way, my main point was/is that Hunspell is not a drop-in "everything will be better" kind of replacement, and that while it may improve some use cases/languages, it also (IMHO seriously) regresses some other use cases.

Your feedback is really valuable, and I think we need to rethink what to do here. Thank you.

**Comment by ctrlcctrlv** [comment]
Please don't forget `:spellrare` if this is going to be the path taken.

**Comment by mcepl** [comment]
> Please don't forget `:spellrare` if this is going to be the path taken.

What does “a rare word” means exactly? Does hunspell support it (or any other spellchecker aside from vimspell)?

**Comment by ctrlcctrlv** [comment]
@mcepl A rare word is…a rare word. I don't know if hunspell supports it, doubt it. It gives you the ability to add a word like let's say `wunderbar`. I'd mark that rare, so I know, when I'm editing a normal document including German words, that that might be (and likely is) a typo. But it wasn't in one document where that was the part of a quote (i.e. _sic_).

**Comment by ctrlcctrlv** [comment]
Another use case of the ability to add rare words is you can add words like `pubic` as rare. I've done that LOL as I rarely want to write `pubic`, it's a typo from `public`. Again, a rare word.

You can also import whole dictionaries as rare. I have all of `~/.local/share/nvim/plugged/medical-spell-files/spell/medical_en` marked rare as it includes many acronyms that are probably typos.

**Comment by carlosala** [comment]
Is this still planned for v0.10?
It'd be amazing😍

**Comment by ctrlcctrlv** [comment]
I hope not, as still no `:spellrare` :joy:

**Comment by mcepl** [comment]
Would it be too much to ask to rebase on the top of the current master, please? The base of this PR is getting so old that it breaks half of my configuration.

**Comment by vigoux** [comment]
For the record, I am not using this PR anymore.

The amount of work needed to switch to another spelling system made this option take too much time to maintain on my end.

Furthermore, I am not sure we _really want to_ make the switch, e.g. because of `spellrare` and other extremelly niche (yet used) features that vimspell has.

**Comment by justinmk** [comment]
> Furthermore, I am not sure we _really want to_ make the switch, e.g. because of `spellrare` and other extremelly niche (yet used) features that vimspell has.

Is that because of maintaining the bindings and build steps? Or something else?

I see the main blocker as hunspell's poor English suggestions. If that improved, we could drop spellrare.

**Comment by arp242** [comment]
> I see the main blocker as hunspell's poor English suggestions. If that improved, we could drop spellrare.

Also Dutch is much worse, in a way that's similar to English.

I don't know any other languages well enough to have an informed opinion; clearly it improves things for *some* languages, but I have no idea if it's "improves for 10 languages, degrades for 90 languages", "improves for 50 languages, degrades for 50 languages", or "improves for 90 languages, degrades for 10 languages".

It *may* be the case that hunspell degrades things for the majority of users, while improving things for *some* users. It's an open question whether this is actually the case, but it's certainly about more than only English.

**Comment by justinmk** [comment]
Ok, let's close this. Can revisit if the landscape changes.

# Resolution

- Path forward is defined in: https://github.com/neovim/neovim/pull/20131#pullrequestreview-1103362634
- Tracking issue: https://github.com/neovim/neovim/issues/12064
